#Перевірка статиcтичних гіпотез. Z-test.



In [79]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats


0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [80]:
data = pd.read_csv('data.csv')

In [81]:
data

,Unnamed: 0,data
0,0,17.499453
1,1,19.662399
2,2,7.182823
3,3,29.841625
4,4,9.239386
...,...,...
95,95,26.244944
96,96,22.282832
97,97,18.697188
98,98,27.041189


Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [82]:
data = data.data.values
data.shape

(100,)

**Завдання 1**. (7 балів)
Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

In [83]:
sample_mean =  np.mean(data)
population_mean = 20
population_std = np.std(data, ddof=1)
sample_size = len(data)
alpha = 0.05

ГІПОТЕЗА:
============================================================


H₀: μ = 20 ($ продажі в середньому на чек не виросли)

H₁: μ > 20 ($ продажі в сердньому на чек виросли за липень)

Тип тесту: правосторонній (one-tailed)

Чому саме правосторонній тест?

Ми намагаємося з'ясувати, чи наші продажі за липень виросли , порівняно з середніми в попередні місяця, тому маємо перевірити чи чек вищий, а не просто вдрізняється.


ОБЧИСЛЕННЯ СТАНДАРТНОЇ ПОМИЛКИ:
============================================================
Формула: SE = σ / √n

SE = 6.25356123240434 / √100

SE = 6.25356123240434 / 10

SE = 0.625

Інтерпретація: В середньому, вибіркове середнє відхиляється
від справжнього середнього на ±0.625356123240434 $ за середній чек.

In [65]:
standard_error = population_std / np.sqrt(sample_size)
print(standard_error)
print(population_std)
print(len(data))
print(sample_mean)

0.625356123240434
6.25356123240434
100
19.37755319670441


РОЗРАХУНОК Z-СТАТИСТИКИ:
============================================================
Формула: z = (x̄ - μ₀) / SE

z = (19.37755319670441 - 20) / 0.625

z = -0.6224468033 / 0.625

z = -0.995347738933505 = близько - 0.995

Інтерпретація:
Наше вибіркове середнє відхиляється на - 0.995 стандартних
помилок від гіпотетичного середнього популяції.


In [60]:
z_statistic = (sample_mean - population_mean) / standard_error
z_statistic

np.float64(-0.995347738933505)

============================================================

КРИТИЧНЕ ЗНАЧЕННЯ:

============================================================

Для рівня значущості α = 0.05 (правосторонній тест)

Критичне z-значення = близько 1.644



In [61]:
z_critical = stats.norm.ppf(1 - alpha)
z_critical

np.float64(1.6448536269514722)

============================================================

P-VALUE:

============================================================

p-value = 0.8402164165464012

Що це означає?
Якщо H₀ правильна (сердній чек не виріс), то ймовірність
випадково отримати середній чек  ≥ 19.37 становить 0.8402164165464012
або приблизно 0.84 %

Це НАДЗВИЧАЙНО мало! (менше 1%)

In [62]:
p_value = 1 - stats.norm.cdf(z_statistic)
p_value

np.float64(0.8402164165464012)

============================================================

Порівняння z-статистики з критичним значенням

============================================================

z-статистика = -0.995

Критичне значення = 1.644

Порівняння: -0.995 < 1.644

Висновок: z-статистика менша за критичне значення

Рішення: не ВІДХИЛЯЄМО H₀

============================================================

Порівняння p-value з рівнем значущості

============================================================


p-value = 0.840

Рівень значущості α = 0.05

Порівняння: 0.840 > 0.05

Висновок: p-value більше за рівень значущості

Рішення: не ВІДХИЛЯЄМО H₀

============================================================

ФІНАЛЬНИЙ ВИСНОВОК:

============================================================


Твердження що до статистичного підвищення продажів за липень НЕ ПІДТВЕРДЖУЄТЬСЯ!

**Завдання 2.** (3 бали)
Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

In [84]:
from statsmodels.stats import weightstats as stests

In [85]:
z_stat_sm, p_val_sm = stests.ztest(data,
                                  value=population_mean,
                                  alternative='larger')

print(f"Результати z-тесту:")
print(f"  Z-статистика: {z_stat_sm:.3f}")
print(f"  P-value: {p_val_sm:.6f}")
print()

Результати z-тесту:
  Z-статистика: -0.995
  P-value: 0.840216



ІНТЕРПРЕТАЦІЯ:
----------------------------------------
   НЕ ВІДХИЛЯЄМО H₀

   p-value (0.840) > α (0.05)

   Висновок: при рівні значущості 0.05, ми можемо стверджувати , що нам не вдалося підняти середній чек за липень.

Розрахунки збігаються.